lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,463 @@
+            # See: https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification
+            self._session.verify = self.cert or None
+        return self._session
+
+    def close(self):
+        """Close HTTP session."""
+        if self._session is not None:
+            self._session.close()
+            self._session = None
+
+    @functools.lru_cache(256)
+    def format_label(self, label: str) -> str:
+        """Build label to match prometheus format.
+
+        `Label format <https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels>`_
+        """
+        for char_from, char_to in self.label_replace_with:
+            label = label.replace(char_from, char_to)
+        return "".join(char for char in label if char in self.label_allowed_chars)
+
+    def build_tags(self, record: logging.LogRecord) -> Dict[str, Any]:
+        """Return tags that must be send to Loki with a log record."""
+        tags = dict(self.tags) if isinstance(self.tags, ConvertingDict) else self.tags
+        tags = cast(Dict[str, Any], copy.deepcopy(tags))
+        tags[self.level_tag] = record.levelname.lower()
+        tags[self.logger_tag] = record.name
+
+        extra_tags = getattr(record, "tags", {})
+        if not isinstance(extra_tags, dict):
+            return tags
+
+        for tag_name, tag_value in extra_tags.items():
+            cleared_name = self.format_label(tag_name)
+            if cleared_name:
+                tags[cleared_name] = tag_value
+
+        return tags
+
+
+class LokiHandler(logging.Handler):
+    """Log handler that sends log records to Loki.
+
+    `Loki API <https://github.com/grafana/loki/blob/mas<#NOWOKE>ter/docs/api.md>`
+    """
+
+    def __init__(
+        self,
+        url: str,
+        tags: Optional[dict] = None,
+        # username, password tuple
+        cert: Optional[str] = None,
+    ):
+        """Create new Loki logging handler.
+
+        Arguments:
+            url: Endpoint used to send log entries to Loki (e.g. `https://my-loki-instance/loki/api/v1/push`).
+            tags: Default tags added to every log record.
+
+            # FIXME: Session expects a .pem file it says
+            cert: Optional absolute path to cert file for TLS auth.
+
+        """
+        super().__init__()
+        self.emitter = LokiEmitter(url, tags, cert)
+
+    def handleError(self, record):  # noqa: N802
+        """Close emitter and let default handler take actions on error."""
+        self.emitter.close()
+        super().handleError(record)
+
+    def emit(self, record: logging.LogRecord):
+        """Send log record to Loki."""
+        # noinspection PyBroadException
+        try:
+            self.emitter(record, self.format(record))
+        except Exception:
+            self.handleError(record)
+
+
+def is_enabled() -> bool:
+    """Whether charm logging is enabled."""
+    return os.getenv(CHARM_LOGGING_ENABLED, "1") == "1"
```

Comment:
Charms have bursts of logs on event hooks. Should this be opt-in rather than opt-out?

Reply:
same goes for traces. what is the concern exactly? 
It's an ephemeral process so it will naturally have activity when it runs otherwise no activity.
If it's opt-in, who would opt-in?